src/common/rate_limit/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+        match value {
+            None => Self::Disabled,
+            Some(0) => Self::Pause,
+            Some(rate) => Self::Fixed(unsafe { NonZeroU64::new_unchecked(rate as _) }),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Check {
+    Ok,
+    Retry(Duration),
+    RetryAfter(oneshot::Receiver<()>),
+}
+
+impl Check {
+    pub fn is_ok(&self) -> bool {
+        matches!(self, Self::Ok)
+    }
+}
+
+/// Shared behavior for rate limiters.
+pub trait RateLimiterTrait: Send + Sync + 'static {
+    /// Return current throttle policy.
+    fn rate_limit(&self) -> RateLimit;
+
+    /// Check if the request with the given quota is supposed to be allowed at the moment.
+    ///
+    /// On success, the quota will be consumed. [`Check::Ok`] is returned.
+    /// The caller is supposed to proceed the request with the given quota.
+    ///
+    /// On failure, [`Check::Retry`] or [`Check::RetryAfter`] is returned.
+    /// The caller is supposed to retry the check after the given duration or retry after receiving the signal.
+    fn check(&self, quota: u64) -> Check;
+}
+
+/// A rate limiter that supports multiple rate limit policy and online policy switch.
+pub struct RateLimiter {
+    inner: ArcSwap<Box<dyn RateLimiterTrait>>,
+}
+
+impl RateLimiter {
+    fn new_inner(rate_limit: RateLimit) -> Box<dyn RateLimiterTrait> {
+        match rate_limit {
+            RateLimit::Disabled => Box::new(InfiniteRatelimiter),
+            RateLimit::Fixed(rate) => Box::new(FixedRateLimiter::new(rate)),
+            RateLimit::Pause => Box::new(PausedRateLimiter::default()),
+        }
+    }
+
+    /// Create a new rate limiter with given rate limit policy.
+    pub fn new(rate_limit: RateLimit) -> Self {
+        let inner: Box<dyn RateLimiterTrait> = Self::new_inner(rate_limit);
+        let inner = ArcSwap::new(Arc::new(inner));
+        Self { inner }
+    }
+
+    /// Update rate limit policy of the rate limiter.
+    ///
+    /// Returns the old rate limit policy.
+    pub fn update(&self, rate_limit: RateLimit) -> RateLimit {
+        let old = self.rate_limit();
+        if self.rate_limit() == rate_limit {
+            return old;
+        }
+        let inner = Self::new_inner(rate_limit);
+        self.inner.store(Arc::new(inner));
+        old
+    }
+
+    /// Monitor the rate limiter with related table id.
+    pub fn monitored(self, table_id: impl Into<TableId>) -> MonitoredRateLimiter {
+        let metric = METRICS.with_guarded_label_values(&[&table_id.into().to_string()]);
+        let rate_limit = AtomicU64::new(self.rate_limit().to_u64());
+        MonitoredRateLimiter {
+            inner: self,
+            metric,
+            rate_limit,
+        }
+    }
+
+    pub fn rate_limit(&self) -> RateLimit {
+        self.inner.load().rate_limit()
+    }
+
+    pub fn check(&self, quota: u64) -> Check {
+        self.inner.load().check(quota)
+    }
+
+    pub async fn wait(&self, quota: u64) {
```

Comment:
So for a fixed limit, what if `quota > limit`? Will this method finally return?

Reply:
It equals to wait for at least  `quota as f64 / limit as f64`, rather than returning an `InsufficientCapacity` error.

Let me add an unit test for this case.